tiatoolbox/models/engine/engine_abc.py

```diff
@@ -395,20 +415,253 @@ def setup_patch_dataset(
+            denominator = None
+            output = np.zeros(list(canvas_shape), dtype=np.float32)
+        else:
+            coordinates = output["coordinates"]
+            predictions = output["probabilities"]
+            num_class = np.array(predictions[0]).shape[0]
+            denominator = np.zeros(canvas_shape)
+            output = np.zeros([*list(canvas_shape), num_class], dtype=np.float32)
+
+        for idx, bound in enumerate(coordinates):
+            prediction = predictions[idx]
+            # assumed to be in XY
+            # top-left for output placement
+            tl = np.ceil(np.array(bound[:2]) * fx).astype(np.int32)
+            # bot-right for output placement
+            br = np.ceil(np.array(bound[2:]) * fx).astype(np.int32)
+            output[tl[1] : br[1], tl[0] : br[0]] += prediction
+            if denominator is not None:
+                denominator[tl[1] : br[1], tl[0] : br[0]] += 1
+
+        # deal with overlapping regions
+        if denominator is not None:
+            output = output / (np.expand_dims(denominator, -1) + 1.0e-8)
+            if not return_raw:
+                # convert raw probabilities to predictions
+                if post_proc_func is not None:
+                    output = post_proc_func(output)
+                else:
+                    output = np.argmax(output, axis=-1)
+                # to make sure background is 0 while class will be 1...N
+                output[denominator > 0] += 1
+
+        return output
 
     @abstractmethod
-    def infer_wsi(self: EngineABC) -> NoReturn:
+    def infer_wsi(
+        self: EngineABC,
+        dataloader: torch.utils.data.DataLoader,
+        img_path: Path,
+        img_label: str,
+        highest_input_resolution: list[dict],
+        *,
+        merge_predictions: bool,
+        **kwargs: dict,
+    ) -> list:
         """Model inference on a WSI."""
-        raise NotImplementedError
+        # return coordinates of patches processed within a tile / whole-slide image
+        return_coordinates = True
+
+        cum_output = {
+            "probabilities": [],
+            "predictions": [],
+            "coordinates": [],
+            "labels": [],
+        }
+
+        # get return flags from kwargs or set to False, useful in Annotation Store
+        return_labels = kwargs["return_labels"] if "return_labels" in kwargs else False
+
+        for _, batch_data in enumerate(dataloader):
+            batch_output_probabilities = self.model.infer_batch(
+                self.model,
+                batch_data["image"],
+                device=self.device,
+            )
+            # We get the index of the class with the maximum probability
+            batch_output_predictions = self.model.postproc_func(
+                batch_output_probabilities,
+            )
+
+            # tolist might be very expensive
+            cum_output["probabilities"].extend(batch_output_probabilities.tolist())
```

Comment:
I think these values should be directly written to zarr otherwise it will create an overhead and complexity specially when we extend to more complex algorithms.

Reply:
[Update] This is partially improved. I've added another helper function to convert/persist the ndarrays as zarr arrays directly. 
[Pending] Instead of `.extend()` which eventually converts it to a list of zarr arrays which will again be inefficient. I'm going to have them as a zarr group. But this breaks the existing dictionary to Annotation store code.

Reply:
[Update] Fully improved. I've added the `wsi_batch_output_to_zarr_group()` function in `misc.py` the saves the intermediate batch outputs of TIAToolbox engines directly to a Zarr file
Full diff: [[commit 9344c0639b](https://github.com/TissueImageAnalytics/tiatoolbox/pull/737/commits/9344c0639b30efc40654d2011b677a8ed1d4067d)]